### PR TITLE
Safer template resolver

### LINKED ISSUE #2761: Classpath issues with adding Giter8
1) sbt-android trying to use Javassist, by @pfn
- https://gitter.im/sbt/sbt?at=57ec2e83783fe9b945a59741

2) sbt-scalajs needing to use commons-compress 1.9 but failing because it gets commons-compress 1.6, by @dispalt
- https://gitter.im/sbt/sbt?at=57e9b2caa96e034d3d9a29c5
- https://gitter.im/sbt/sbt?at=57ea9f60d02791686c3f8c82

3) Older versions of sbt-git using older version of JGit and failing, by @SethTisue
- https://github.com/scala/community-builds/pull/290#issuecomment-248458277

### PR DESCRIPTION
**Pull request by eed3si9n**
Fixes #2761
Ref #2705

With sbt 0.13.13-RC1 rediscovered that the dependency pulled in from
Giter8 was affecting the plugins. To avoid this, this change splits up
the template resolver implementation to another module called
sbt-giter8-resolver, and it will be downloaded using Ivy into
`~/.sbt/0.13/templates/`, and then launched reflectively using Java as
the interface.

We need to wait till https://bintray.com/sbt/maven-releases/sbt-giter8-resolver get published to Maven Central.

### CODE REVIEW

File: main/src/main/scala/sbt/TemplateCommand.scala [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+  private def call(interfaceClassName: String, methodName: String, loader: ClassLoader)(argTypes: Class[_]*)(args: AnyRef*): AnyRef =
+    {
+      val interfaceClass = getInterfaceClass(interfaceClassName, loader)
+      val interface = interfaceClass.newInstance.asInstanceOf[AnyRef]
+      val method = interfaceClass.getMethod(methodName, argTypes: _*)
+      try { method.invoke(interface, args: _*) }
+      catch {
+        case e: java.lang.reflect.InvocationTargetException =>
+          e.getCause match {
+            case t => throw t
```
Comment (dwijnand) [inline main/src/main/scala/sbt/TemplateCommand.scala]: Why the match?

File: main/src/main/scala/sbt/TemplateCommand.scala [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+      val updateUtil = new UpdateUtil(ivyConf, log)
+      val templatesBaseDirectory = new File(globalBase, "templates")
+      val templateId = s"${info.module.organization}_${info.module.name}_${info.module.revision}"
+      val templateDirectory = new File(templatesBaseDirectory, templateId)
+      def jars = (templateDirectory ** -DirectoryFilter).get
+      if (!(info.module.revision endsWith "-SNAPSHOT") && jars.nonEmpty) jars.toList
+      else {
+        IO.createDirectory(templateDirectory)
+        val m = updateUtil.getModule(info.module.copy(configurations = Some("component")), ivyScala)
+        val xs = updateUtil.update(m, templateDirectory)(_ => true) match {
```
Comment (dwijnand) [inline main/src/main/scala/sbt/TemplateCommand.scala]: instead of matching you can `.toList.flatten` here

File: build.sbt [hunk 3]
```diff
@@ -445,11 +445,11 @@ lazy val actionsProj = (project in mainPath / "actions").
...
 // General command support and core commands not specific to a build system
 lazy val commandProj = (project in mainPath / "command").
-  dependsOn(interfaceProj, ioProj, logProj, completeProj, classpathProj, crossProj).
+  dependsOn(interfaceProj, ioProj, logProj, completeProj, classpathProj, crossProj, ivyProj).
   settings(
     testedBaseSettings,
     name := "Command",
-    libraryDependencies ++= Seq(launcherInterface, templateResolverApi, giter8),
+    libraryDependencies ++= Seq(launcherInterface, templateResolverApi),
     dependencyOverrides += plexusUtils
```
Comment (dwijnand) [inline build.sbt]: Can we remove the 2 instance of `dependencyOverrides += plexusUtils`?

**Review by dwijnand — CHANGES_REQUESTED** [review]
Few minor bits, but LGTM on paper.

Did you run this through the 3 issues we hit? Or shall I do that tomorrow?

Also, I'd love if you had time tomorrow to explain to me how you pull this off. Such impressed. Wow.

File: build.sbt (hunk 3 above)
Comment (eed3si9n) [reply build.sbt]: Yes.

File: main/src/main/scala/sbt/TemplateCommand.scala (hunk 1 above)
Comment (eed3si9n) [reply main/src/main/scala/sbt/TemplateCommand.scala]: Remnant from the code that I copy pasted from. I should clean it up.

File: main/src/main/scala/sbt/TemplateCommand.scala [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    {
+      val resultObj = call(info.implementationClass, "isDefined", loader)(
+        classOf[Array[String]])(arguments.toArray)
+      resultObj.asInstanceOf[Boolean]
+    }
+  private def runTemplate(info: TemplateResolverInfo, arguments: List[String], loader: ClassLoader): Unit =
+    call(info.implementationClass, "run", loader)(classOf[Array[String]])(arguments.toArray)
+  private def infoLoader(info: TemplateResolverInfo, config: AppConfiguration,
+    ivyConf: IvyConfiguration, globalBase: File, ivyScala: Option[IvyScala], log: Logger): ClassLoader =
+    ClasspathUtilities.toLoader(classpathForInfo(info, ivyConf, globalBase, ivyScala, log), config.provider.loader)
```
Comment (dwijnand) [inline main/src/main/scala/sbt/TemplateCommand.scala]: TIL ClasspathUtilities.toLoader, pretty cool.

**Comment by eed3si9n** [comment]
I just tested @pfn's scenario https://github.com/sbt/sbt/issues/2761#issuecomment-250509014 and seems to be working.

```
$ xsbt "gen-android com.foo Foo" android:ndkJavah
....
[info] Done packaging.
[success] Total time: 11 s, completed Oct 13, 2016 1:07:23 AM
```

File: main/src/main/scala/sbt/TemplateCommand.scala [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+  private def call(interfaceClassName: String, methodName: String, loader: ClassLoader)(argTypes: Class[_]*)(args: AnyRef*): AnyRef =
+    {
+      val interfaceClass = getInterfaceClass(interfaceClassName, loader)
+      val interface = interfaceClass.newInstance.asInstanceOf[AnyRef]
+      val method = interfaceClass.getMethod(methodName, argTypes: _*)
+      method.invoke(interface, args: _*)
```
Comment (dwijnand) [inline main/src/main/scala/sbt/TemplateCommand.scala]: No I think it's correct here to catch InvocationTargetException and throw it's cause. (Sorry)

Comment (eed3si9n) [reply main/src/main/scala/sbt/TemplateCommand.scala]: ok. Fixed it.

**Review by dwijnand — APPROVED** [review]

File: main/src/main/scala/sbt/TemplateCommand.scala (hunk 5 above)
Comment (dwijnand) [reply main/src/main/scala/sbt/TemplateCommand.scala]: Thanks.

File: main/src/main/scala/sbt/TemplateCommand.scala [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+  def templateCommandParser(state: State) =
+    {
+      val p = (token(Space) ~> repsep(StringBasic, token(Space))) | (token(EOF) map { case _ => Nil })
+      val infos = (state get templateResolverInfos) match {
+        case Some(infos) => infos.toList
+        case None        => Nil
+      }
+      val log = state.globalLogging.full
+      val extracted = (Project extract state)
+      val (s2, ivyConf) = extracted.runTask(Keys.ivyConfiguration, state)
```
Comment (dwijnand) [inline main/src/main/scala/sbt/TemplateCommand.scala:25]: Implemented this way causes `target/` trash when calling `new` unfortunately:

```
target
└── streams
    └── $global
        ├── ivyConfiguration
        │   └── $global
        │       └── streams
        │           └── out
        └── projectDescriptors
            └── $global
                └── streams
                    └── out
```

**Pull request merged**